Fix: #502 - Store the URI scheme in the remote config.

grisuthedragon opened the pull request:
The remotes are stored without the scheme in the configuration.
Due to some hard coded https string, this prevents using insecure
registries. The patch enables storing the scheme as additional value
in the configuration. If the scheme is not present, https is used
as before.

## Description of the Pull Request (PR):

The PR closes the bug #502 by storing the scheme in addition to the URI. This enables using local registries without https access. 


### This fixes or addresses the following GitHub issues:

 - Fixes #502

DrDaveD commented:
Failed the check_source CI, look at the details.

DrDaveD commented:
Does this avoid the need for using the `--no-https` option on pull/push/exec/instance start/run/shell?

Assuming that this is acceptable to the other reviewers, I would like to see a CHANGELOG.md entry added at the end of the changed behaviors section for changes since last release.  It would be good to include an e2e test too but I don't think there's any local registry used in any of the tests and so it might be too much trouble to implement.  Maybe instead just unit tests showing that an http URL can be stored and read back.

grisuthedragon commented:
> Does this avoid the need for using the `--no-https` option on pull/push/exec/instance start/run/shell?

If the pull/push/... is done on a properly configured `http` remote, there is no need for the `--no-https`. If the user configured the remote with http only, then it has to work for all operations without specifying the `no-https` again.  If the library URI contains a full hostname, i.e. the configured remote is overridden, then `--no-https` is still required. In case of `remote add` and `remote login` the https prefix was hard-coded and not affected by `--no-https`. 

> Assuming that this is acceptable to the other reviewers, I would like to see a CHANGELOG.md entry added at the end of the changed behaviors section for changes since last release. It would be good to include an e2e test too but I don't think there's any local registry used in any of the tests and so it might be too much trouble to implement. Maybe instead just unit tests showing that an http URL can be stored and read back.

I add the CHANGELOG entry and try to figure out how to create such a test, since I am not that familiar with go. ( So I would be happy about some help)

dtrudg commented:
Just a quick heads up that SingularityCE added http support for remotes via the `--insecure` flag in 3.9.0. Understand that you might want to go a different way, but if CLI compatibility / `remote.yaml` compatibility is important then it might be worth considering. Cheers!

See: https://github.com/sylabs/singularity/pull/245

DrDaveD commented:
Thanks, DT, I had forgotten about that.  I had made a [PR](apptainer/singularity#6219) to pull it in but at that time we were thinking we would not have remotes at all.  It was then forgotten after remotes were retained.

@grisuthedragon How about if you instead re-do that PR?

grisuthedragon commented:
@DrDaveD 
I think, the PR from singularity would work for me and my colleagues as well, with one exception. If a `http` remote is added, like 
```
apptainer remote add localremote http://library.inter/
```
the `insecure` flag should  be set automatically. So adding a `http` remote implies `insecure` probably with displaying a warning when it is added. Since we live in times, where everybody should be clear about what http means in the sense of transferring passwords and tokens in a public environment.  For compatibility with singluarity, I see that the `--insecure` command line flag should be added as well. 

One can additionally think of `--insecure` allows `https` connections with self-signed or outdated SSL certs.

dtrudg commented:
> Since we live in times, where everybody should be clear about what http means in the sense of transferring passwords and tokens in a public environment. For compatibility with singularity, I see that the `--insecure` command line flag should be added as well.

As a counterpoint... we explicitly chose not to accept `http://` hosts without an `--insecure` flag, as we _don't_ think it's well known to the general user base what this means. Arguably `https` is so pervasive across the web these days, that people are less aware of http vs https.

> One can additionally think of `--insecure` allows `https` connections with self-signed or outdated SSL certs. But looking at [sylabs/singularity#245](https://github.com/sylabs/singularity/pull/245) the definition of `insecure` introduced by @dtrudg is in fact "useHttp", which differs somehow a bit from the way I think of `insecure`. So self-signed SSLs certs are `insecure` in some sense.

See - https://github.com/sylabs/singularity/issues/683 - for a bit context on the SingularityCE side on this.

In some of the comments there we're talking about deprecating the `http-only` flag we have in some places... and moving to an `--insecure` handling across Singularity that behaves more like Docker... where insecure tries https (allowing out of date / self-signed certs), and falls back to http.

This hasn't really gone anywhere yet, mainly due to it needing to consider consistency across the various libraries we use to pull/push things - where we aren't always in control of the exact behavior.

grisuthedragon commented:
@dtrudg 
For now, I will try to implement the `insecure` behaviour of singularity to apptainer (with my automatic http means insecure addon). I will follow the discussion at  https://github.com/sylabs/singularity/issues/683 to see what's going on. (We have the problem with selft signed docker registries as well but there we have a local CA distributed to all machines, but that solution is outdated at many other points)

grisuthedragon commented:
Close due to new approach following the singularity command line interface and behaviour.
